src/ram/transform/MakeIndex.cpp

```diff
@@ -38,131 +38,225 @@
+            if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
+                element = rhs->getElement();
+                return {std::make_unique<RamUndefValue>(), clone(lhs)};
+            }
+        }
+    }
+    return {std::make_unique<RamUndefValue>(), std::make_unique<RamUndefValue>()};
+}
+
+ExpressionPair MakeIndexTransformer::getFloatExpressionPair(
+        const RamConstraint* binRelOp, size_t& element, int identifier) {
+    if (isLessEqualFloat(binRelOp->getOperator())) {
+        // Tuple[level, element] <= <expr>
+        if (const auto* lhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getLHS())) {
+            const RamExpression* rhs = &binRelOp->getRHS();
+            if (lhs->getTupleId() == identifier && rla->getLevel(rhs) < identifier) {
+                element = lhs->getElement();
+                return {std::make_unique<RamUndefValue>(), clone(rhs)};
+            }
+        }
+        // <expr> <= Tuple[level, element]
+        if (const auto* rhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getRHS())) {
+            const RamExpression* lhs = &binRelOp->getLHS();
+            if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
+                element = rhs->getElement();
+                return {clone(lhs), std::make_unique<RamUndefValue>()};
             }
         }
+    }
 
-        if (isGreaterThanConstraint(binRelOp->getOperator())) {
-            // Tuple[level, element] > <expr>
-            // is equivalent to
-            // Tuple[level, element] >= <expr> + 1
+    if (isGreaterEqualFloat(binRelOp->getOperator())) {
+        // Tuple[level, element] >= <expr>
+        if (const auto* lhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getLHS())) {
+            const RamExpression* rhs = &binRelOp->getRHS();
+            if (lhs->getTupleId() == identifier && rla->getLevel(rhs) < identifier) {
+                element = lhs->getElement();
+                return {clone(rhs), std::make_unique<RamUndefValue>()};
+            }
+        }
+        // <expr> >= Tuple[level, element]
+        if (const auto* rhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getRHS())) {
+            const RamExpression* lhs = &binRelOp->getLHS();
+            if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
+                element = rhs->getElement();
+                return {std::make_unique<RamUndefValue>(), clone(lhs)};
+            }
+        }
+    }
+    return {std::make_unique<RamUndefValue>(), std::make_unique<RamUndefValue>()};
+}
+
+// Retrieves the <expr1> <= Tuple[level, element] <= <expr2> part of the constraint as a pair { <expr1>,
+// <expr2> }
+ExpressionPair MakeIndexTransformer::getLowerUpperExpression(
+        RamCondition* c, size_t& element, int identifier) {
+    if (auto* binRelOp = dynamic_cast<RamConstraint*>(c)) {
+        // TODO: FIXME: how does this interact w/ `FEQ`?
```

Comment:
Is this FIXME addressed by the float check below?

Reply:
There's an important adjustment I need to make in constructPattern() to use the appropriate type of BinaryConstraintOp which I will work on soon. It's my birthday today so I will realistically finish this tomorrow!

Reply:
No problem, just checking the comment was still accurate.